lightning/src/util/chacha20.rs

```diff
@@ -151,6 +151,17 @@ mod real_chacha {
 			ChaCha20{ state: ChaCha20::expand(key, nonce), output: [0u8; BLOCK_SIZE], offset: 64 }
 		}
 
+		/// Get one block from a ChaCha stream. Panics if `key`'s len is not 32 bytes or `nonce`'s len
+		/// is not 16 bytes.
+		pub fn get_single_block(key: &[u8], nonce: &[u8]) -> [u8; 32] {
```

Comment:
ChannelManager's kinda big so i kinda like shifting code out of it where possible. It is weird that the API is called `get_single_block` when `BLOCK_SIZE` is 64. Could rename to `get_32byte_block` :grimacing:

Reply:
Sure, would you be happy with (in channelmanager.rs): 
```
impl InboundPayment {
  fn create(..) -> Result<(PaymentSecret, PaymentHash), ..>
  fn create_from_hash(..) -> Result<PaymentSecret, ..> ..
  fn verify
  fn get_encryption_block
}
```
and maybe some other DRY'ing methods.
@TheBlueMatt are you fine with shifting `create_inbound_payment` and such logic into its own struct + moving `ChaCha20::get_single_block` in there?

Reply:
I mean this is strictly a cryptographic utility method, though? Seems weird to have it in `channelmanager`, which doesn't implement any crypto itself. What if we change the parameters here to not be slices but instead be array references? I think the callsites are fine with that and then we don't need stupid slice-length assertions.

Reply:
Using array references would help, yes, but the particular array lengths are for a specific use case. And this method isn't doing anything much other than initializing and calling a method on `ChaCha20` for that specific use case.

Reply:
It doesn't just call methods, though, it calls specific private methods and tweaks the internal ChaCha state in a way which isn't compatible with the general ChaCha stream API. Note that there's ~no difference between returning the first 32 bytes of the block (as done now) and returning 64 bytes - we calculate the whole thing anyway so its just a bit more memcpy, or none if we add an `#[inline]`.

Reply:
Hmm... I see. I guessing calling `ChaCha20::new` wouldn't work because the asserts in there. Suppose using array references seems like a reasonable approach then.

Reply:
@jkczyz @TheBlueMatt acks on this approach + the slice->array references for ChaCha? https://github.com/lightningdevkit/rust-lightning/pull/1177#discussion_r764244269

Reply:
I don't think we want to mix payment logic in with chacha cryptography logic. Lets keep the get-single-block method (and maybe adapt it to return a block and take array references). If we want to refactor the inbound payment methods out from channelmanager into their own file/module/struct I guess that's fine, though given the struct isn't *stored* anywhere it seems strange to make it a struct vs just some helper functions (ie the functions that already exist just moved elsewhere).

Reply:
That said, I'm not convinced it needs to be moved elsewhere either, but if we want to `onion_utils` is pretty bare now.

Reply:
> I don't think we want to mix payment logic in with chacha cryptography logic. Lets keep the get-single-block method (and maybe adapt it to return a block and take array references).

SGTM

> If we want to refactor the inbound payment methods out from channelmanager into their own file/module/struct I guess that's fine, though given the struct isn't stored anywhere it seems strange to make it a struct vs just some helper functions (ie the functions that already exist just moved elsewhere).

To me this is more about avoiding the accumulation of organizational debt, for lack of a better term. `ChannelManager` seems to accumulate a lot of code, so keeping logically dependent code together (whether in the same file or elsewhere) is a way of avoiding this. Using an abstraction makes it clear to future authors that it belongs together without it getting lost in the shuffle of other methods that may be added in the future. Putting utilities near their caller (e.g. in the case of `verify_inbound_payment_data`) goes against this for not much benefit -- we try to use good names so we don't need to look at the code that it is abstracting away.

Reply:
> ChannelManager seems to accumulate a lot of code, so keeping logically dependent code together (whether in the same file or elsewhere) is a way of avoiding this.

Yea, that's totally fair.

Reply:
Added new `inbound_payment` module with helpers